### PR TITLE
LLM Finetuning Toolkit

### PR DESCRIPTION
**Pull request by benjaminye**
This PR introduces a config-driven LLM fine-tuning experience.

**Key Features**
- Launch fine-tuning by defining a single `config.yaml` file
- Interactive and helpful print outputs
- Experiment checkpointing that saves and loads experiment artefacts at various stages, enabling one to pickup where left off

**Key Modules**
- `DatasetGenerator`: loads model into a huggingface `Dataset` object based on file extension (csv, json, or huggingface repo)
- `ModelLoader`: manages loading of base model, appending lora weights, and merging lora weights
- `InferenceRunner`: runs merged model on test set, outputs live streaming predictions, and saves predictions into csv

**Road Map**
- [ ] adopt `poetry` for dependency management
- [ ] create a docker container for predictable deployment
- [ ] multi-GPU support
- [ ] stratified sampling for dataset
- [ ] ablation studies
- [ ] documentation

### CODE REVIEW

**Comment by RohitSaha** [comment]
@benjaminye Aren't we doing stratified sampling now?

File: toolkit/config.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    >- # prompt, make sure column inputs are enclosed in {} brackets and that they match your data
+    Below is an instruction that describes a task. 
+    Write a response that appropriately completes the request. 
+    ### Instruction: {instruction}
+    ### Input: {input}
+    ### Output:
+  prompt_stub:
+    >- # Stub to add for training at the end of prompt, for test set or inference, this is omitted; make sure only one variable is present
+    {output}
+  test_size: 10 # Proportion of test as % of total; if integer then # of samples
```
Comment (RohitSaha) [inline toolkit/config.yml:17]: How to differentiate between proportion and # of samples?

File: toolkit/config.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+model:
+  hf_model_ckpt: "NousResearch/Llama-2-7b-hf"
+  quantize: true
+  bitsandbytes:
+    load_in_4bit: true
+    bnb_4bit_compute_dtype: "bf16"
+    bnb_4bit_quant_type: "nf4"
+
+# LoRA Params -------------------
+lora:
```
Comment (RohitSaha) [inline toolkit/config.yml:30]: Did we include neftune functionality?

File: toolkit/config.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    lr_scheduler_type: "constant"
+  sft_args:
+    max_seq_length: 1024
+    # neftune_noise_alpha: None
+
+inference:
+  max_new_tokens: 200
+  use_cache: True
+  do_sample: True
+  top_p: 0.6
```
Comment (RohitSaha) [inline toolkit/config.yml:67]: let's set top_p to 0.9, and temperature to 0.1 as default values

File: toolkit/config.yml (hunk 1 above)
Comment (benjaminye) [reply toolkit/config.yml:17]: This is uniquely determined by the data type passed in.

**proportion**: float;
**\# of samples**: int;

this is built into huggingface `dataset`'s [`train_test_split()`](https://huggingface.co/docs/datasets/v2.15.0/en/package_reference/main_classes#datasets.Dataset.train_test_split) method.

**Comment by RohitSaha** [comment]
@benjaminye Let's ensure the packages we use **_only_** contain permissive licenses, i.e, Apache 2.0 and MIT. We should remove any dependencies that have copyleft license.

**Comment by benjaminye** [comment]
> @benjaminye Aren't we doing stratified sampling now?

Unfortunately this isn't as simple. I used huggingface's recommended [`from_generator()`](https://huggingface.co/docs/datasets/v2.15.0/en/create_dataset#from-local-files) method to read in files directly into a `Dataset` object; instead of loading the files into a pandas DataFrame first (which isn't as memory & compute efficient).

Not loading into `DataFrame` precludes us from using `scikit-learn`'s `train_test_split()` method; instead we have to use the method from huggingface `dataset`. However, that method expects the `stratify_by` column to be  a [`ClassLabel`](https://huggingface.co/docs/datasets/v2.15.0/en/package_reference/main_classes#datasets.ClassLabel) (which is an integer representation). A hacky method is to cast it to `ClassLabel` and then cast it back into string `Value`, an avenue that I'm exploring.

Moreover, since stratified sampling only applies to classification tasks, I decided to think about this a bit more and release it in the next update.

**Comment by benjaminye** [comment]
> @benjaminye Let's ensure the packages we use **_only_** contain permissive licenses, i.e, Apache 2.0 and MIT. We should remove any dependencies that have copyleft license.

I've checked all the new dependencies and they all have permissive licenses.

File: toolkit/config.yml (hunk 3 above)
Comment (benjaminye) [reply toolkit/config.yml]: resolved at commit [161abc5](https://github.com/georgian-io/LLM-Finetuning-Hub/pull/67/commits/161abc5b81891c514a40044202a587dc8e82b9a0)

**Comment by RohitSaha** [comment]
Thanks for making the changes!

**Review by RohitSaha — APPROVED** [review]
